runtime-sdk-macros/src/evm_derive.rs

```diff
@@ -0,0 +1,467 @@
+            FnArg::Receiver(recv) => {
+                recv.self_token
+                    .span
+                    .unwrap()
+                    .error("Dispatchable methods can't have a receiver")
+                    .emit();
+                return None;
+            }
+            FnArg::Typed(typed) => match *typed.pat {
+                Pat::Ident(_) => continue,
+                _ => {
+                    typed
+                        .colon_token
+                        .span
+                        .unwrap()
+                        .error("Unsupported function input type")
+                        .emit();
+                    return None;
+                }
+            },
+        }
+    }
+    if signature.convert && signature.args.len() != f.sig.inputs.len() - 1 {
+        signature
+            .name
+            .span()
+            .unwrap()
+            .error("Number of function arguments doesn't match EVM signature")
+            .emit();
+        return None;
+    }
+
+    let byte_tokens = signature.selector.iter().map(|&b| quote!(#b));
+    let byte_list = quote!(#(#byte_tokens ,)*);
+    Some((signature, attr, byte_list))
+}
+
+fn generate_method_call(attr: &EvmMethod, sig: &Signature) -> Option<TokenStream> {
+    let mut decoder_list: Vec<TokenStream> = Vec::new();
+    let mut call_args: Vec<Ident> = Vec::new();
+    let mut arg_decls: Vec<TokenStream> = Vec::new();
+    for (i, evm_ty) in sig.args.iter().enumerate() {
+        let arg_name = format_ident!("arg_{}", i);
+        call_args.push(arg_name.clone());
+        match evm_ty.as_str() {
+            "address" => {
+                decoder_list.push(quote!(::ethabi::ParamType::Address));
+                arg_decls.push(quote! {
+                    let #arg_name = decoded_args[#i].clone().into_address().unwrap();
+                });
+            }
+            "uint256" => {
+                decoder_list.push(quote!(::ethabi::ParamType::Uint(256)));
+                let temp_name = format_ident!("{}_uint", arg_name);
+                arg_decls.push(quote! {
+                    let #temp_name = decoded_args[#i].clone().into_uint().unwrap();
+                    if #temp_name.bits() > (u128::BITS as usize) {
+                        return Some(Err(::evm::executor::stack::PrecompileFailure::Error {
+                            exit_status: ::evm::ExitError::Other("integer overflow".into()),
+                        }));
+                    }
+                    let #arg_name = #temp_name.as_u128();
+                });
+            }
+            ty => {
+                attr.signature
+                    .span()
+                    .unwrap()
+                    .error(format!("Unknown argument type '{}'", ty))
+                    .emit();
+            }
+        }
+    }
+    let method_name = sig.name.clone();
+    Some(quote! { {
+        let decoded_args = match ::ethabi::decode(&[#(#decoder_list),*], &handle.input()[#SELECTOR_LENGTH..]) {
```

Comment:
Should this use `decode_whole` instead? What is standard EVM behavior in this case as we should probably be consistent?

Reply:
If I'm reading the specification (https://docs.soliditylang.org/en/v0.8.28/abi-spec.html) right, it doesn't specify decoding any further than "inverse of encoding", so various decoding corner cases are not covered. In practice apparently neither under- nor overflows of call data are handled specially, so various attacks are possible...